### PR TITLE
[Impeller] stencil buffer record/replay instead of MSAA storage.

### PR DESCRIPTION
**Pull request by jonahwilliams**
When restoring from a backdrop filter, replay the clip affecting cmds into a new stencil buffer instead of storing 4x MSAA stencil buffer.

Fixes https://github.com/flutter/flutter/issues/137561
Fixes https://github.com/flutter/flutter/issues/137448
Fixes https://github.com/flutter/flutter/issues/137302

Helps https://github.com/flutter/flutter/issues/137108

### CODE REVIEW

**Comment by jonahwilliams** [comment]
I _think_ (hope) we have golden test coverage for this, right? 😄

**Comment by jonahwilliams** [comment]
Currently investigating performance on iOS using the backdrop filter perf (Partial repaint disabled). With this change:

```
  "average_gpu_frame_time": 247.9492189,
  "90th_percentile_gpu_frame_time": 249.023438,
  "99th_percentile_gpu_frame_time": 249.023438,
  "worst_gpu_frame_time": 249.023438,
  ...
   "average_frame_rasterizer_time_millis": 225.71757142857146,
  "stddev_frame_rasterizer_time_millis": 0.623428571428572,
  "90th_percentile_frame_rasterizer_time_millis": 226.62,
  "99th_percentile_frame_rasterizer_time_millis": 227.661,
  "worst_frame_rasterizer_time_millis": 227.661,
  ```
  
  Without it
  
  ```
    "average_gpu_frame_time": 307.61718740000003,
  "90th_percentile_gpu_frame_time": 310.058594,
  "99th_percentile_gpu_frame_time": 310.058594,
  "worst_gpu_frame_time": 310.058594,
  
  ...
  
    "average_frame_rasterizer_time_millis": 279.90400000000005,
  "stddev_frame_rasterizer_time_millis": 1.0763076923076982,
  "90th_percentile_frame_rasterizer_time_millis": 281.933,
  "99th_percentile_frame_rasterizer_time_millis": 282.053,
  "worst_frame_rasterizer_time_millis": 282.053,
  ```
  
So measurably better!

File: impeller/entity/entity_pass.h [hunk 1]
```diff
@@ -309,4 +312,24 @@ class EntityPass {
   EntityPass& operator=(const EntityPass&) = delete;
 };
 
+/// @brief A class that tracks all clips that have been recorded in the current
+///        entity pass stencil.
+///
+///        These clips are replayed when restoring the backdrop so that the
+///        stencil buffer is left in an identical state.
+class EntityPassClipReplay {
```
Comment (bdero) [inline impeller/entity/entity_pass.h:320]: `EntityPassClipRecorder`?

File: impeller/entity/entity_pass.cc [hunk 2]
```diff
@@ -285,8 +283,8 @@ static EntityPassTarget CreateRenderTarget(ContentContext& renderer,
             .resolve_storage_mode = StorageMode::kDevicePrivate,
             .load_action = LoadAction::kDontCare,
             .store_action = StoreAction::kMultisampleResolve,
-            .clear_color = clear_color},   // color_attachment_config
-        GetDefaultStencilConfig(readable)  // stencil_attachment_config
+            .clear_color = clear_color},  // color_attachment_config
+        kDefaultStencilConfig             // stencil_attachment_config
```
Comment (bdero) [inline impeller/entity/entity_pass.cc:287]: Should be able to remove the readable param from CreateRenderTarget now

File: impeller/entity/inline_pass_context.h [hunk 3]
```diff
@@ -40,7 +40,7 @@ class InlinePassContext {
   std::shared_ptr<CommandBuffer> command_buffer_;
   std::shared_ptr<RenderPass> pass_;
   uint32_t pass_count_ = 0;
-  uint32_t total_pass_reads_ = 0;
```
Comment (bdero) [inline impeller/entity/inline_pass_context.h:43]: Good riddance...

**Review by bdero — APPROVED** [review]
Just minor comments.

I really hope this doesn't generally regress Metal in the common case. I'd much rather have this recorder/replay mechanism than the `total_pass_reads_` mess.

File: impeller/entity/entity_pass.h (hunk 1 above)
Comment (jonahwilliams) [reply impeller/entity/entity_pass.h:320]: That is much better

File: impeller/entity/entity_pass.cc (hunk 2 above)
Comment (jonahwilliams) [reply impeller/entity/entity_pass.cc:287]: Done

**Comment by jonahwilliams** [comment]
Alright lets see this number go down:

https://flutter-flutter-perf.skia.org/e/?queries=device_type%3DPixel_7_Pro%26sub_result%3Daverage_frame_rasterizer_time_millis%26sub_result%3Daverage_gpu_frame_time%26test%3Danimated_advanced_blend_perf__timeline_summary%26test%3Danimated_advanced_blend_perf_opengles__timeline_summary&selected=commit%3D37731%26name%3D%252Carch%253Dintel%252Cbranch%253Dmaster%252Cconfig%253Ddefault%252Cdevice_type%253DPixel_7_Pro%252Cdevice_version%253Dnone%252Chost_type%253Dlinux%252Csub_result%253Daverage_frame_rasterizer_time_millis%252Ctest%253Danimated_advanced_blend_perf_opengles__timeline_summary%252C

**Pull request merged**